[Proxy]: add `healthz` verification when determining HC response for eTP:Local

alexanderConstantinescu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

/kind bug 

#### What this PR does / why we need it:

This PR adds an accessor field for `service_health.go` (the one answering to LB HCs for eTP:Local) to the `healthz` state computed in `proxier_health.go`. It will then return a HTTP 200 is the service has a ready endpoint on the node AND if `healthz` is true.   

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #[115825](https://github.com/kubernetes/kubernetes/issues/115825)

Kube-proxy will now include the `healthz` state in its response to the LB HC as to avoid indicating to the LB that it should use the node in question when Kube-proxy is not healthy.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Fixes #115825. Kube-proxy will now include the `healthz` state in its response to the LB HC as to avoid indicating to the LB that it should use the node in question when Kube-proxy is not healthy.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

/sig network
/assign @thockin

aojea commented:
/ok-to-test
/assign @thockin @danwinship 

We've found recently that the node handlers are feature gated, if we want to use this handlers we should remove the feature gate check on the Node informer https://github.com/kubernetes/kubernetes/pull/111344#discussion_r936833738


https://github.com/kubernetes/kubernetes/blob/d96af5f2761c23eddde206aee7b7147a9692b737/cmd/kube-proxy/app/server.go#L753-L766

aojea commented on pkg/proxy/healthcheck/service_health.go:
```diff
@@ -36,6 +36,12 @@ import (
 )
 
+const (
+	// ToBeDeletedTaint is a taint used by the CLuster Autoscaler before marking a node for deletion. Defined in
+	// https://github.com/kubernetes/autoscaler/blob/e80ab518340f88f364fe3ef063f8303755125971/cluster-autoscaler/utils/deletetaint/delete.go#L36
+	ToBeDeletedTaint = "ToBeDeletedByClusterAutoscaler"
```
this can be a slippery slope if we have to support all the projects that add taints on nodes :/

alexanderConstantinescu replied:
Hmm, I wanted to start off by saying: "I can't say anything about the future for the proxy code, but the CCM has been looking at only this one taint on services for...", and while looking at the service controller code and inspecting who added this taint, it seems it was added with: https://github.com/kubernetes/kubernetes/pull/105946 which had a very dubious reason for adding it in the first place:

> Prior to https://github.com/kubernetes/kubernetes/pull/90823, these nodes would get removed from the LoadBalancer backend once they were marked as Unschedulable. Now that Unschedulable nodes are tolerated by the node predicate, these nodes will only get removed when the nodes have been deleted.  

Which is not true _today_, the unschedulable predicate was re-added by me in the CCM...but now I am asking myself questions as to whether that was good, the PR removing the unschedulability predicate (#90823) mentioned a production outage

> This caused a production outage when nodes in an autoscaling group were marked unschedulable so the autoscaler would gracefully terminate them - instead the ingress controllers on those nodes were immediately removed from the LB pool and no new ingress requests could be accepted.

/cc @thockin

alexanderConstantinescu replied:
I've revisited this condition as a part of my work on https://github.com/kubernetes/kubernetes/pull/111658. I think we can remove this condition. It didn't have any clear reasons for being added in the first place and logically, today, it doesn't make sense. 

We can discuss this on the on-going KEP I have: https://github.com/kubernetes/enhancements/pull/3460 now during the 1.26 KEP cycle. I thought upon the life cycle of the Node and its relation to the LB in that document. I think we can remove this condition altogether at this point, but I will leave the PR as-is until we get some clarity and consensus on the KEP. Once done, I will update the PR for 1.26.

aojea commented on pkg/proxy/healthcheck/service_health.go:
```diff
@@ -75,14 +85,20 @@ func newServiceHealthServer(hostname string, recorder events.EventRecorder, list
+	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses, healthzServer)
+}
+
+type nodeState struct {
+	isHealthy bool
```
don't we have to do this variable thread safe, it is accessed from multiple places
use atomic perhaps?

alexanderConstantinescu replied:
Good point, we probably need to "wrap" `SyncNode` in 

```
hcs.lock.Lock()
defer hcs.lock.Unlock()
```

As has been done for `SyncService`. The read should already be protected though in `ServeHTTP` by the read lock.

aojea replied:
I like this atomic bool vars for these cases, your choice though

https://github.com/kubernetes/kubernetes/blob/ee505677f70439ff8f5515cc899eab7c18cdb1dd/pkg/proxy/iptables/proxier.go#L588-L598

https://gist.github.com/mahan/6256149

alexanderConstantinescu replied:
But in this case there really is only one writer, there's no concurrency. `SyncNode` is only invoked by the node informer, and its event handlers (add/update/delete) cannot be called concurrently. So we hence only ever have one writer at any given moment in time. 

Making it an atomic bool would give the impression to people reading the code that we are dealing with concurrency, where in fact that is not the case.

alexanderConstantinescu replied:
Nevermind what I said above, I was confused. There is concurrency, but I find the code reads better by using `hcs.lock`

aojea commented on pkg/proxy/healthcheck/proxier_health.go:
```diff
@@ -91,6 +97,33 @@ func (hs *proxierHealthServer) QueuedUpdate() {
+// performance w.r.t updating the service rules on the node, specifically:
+// it's healthy when it's fully synced and not timing out syncing the
+// service rules.
+func (hs *proxierHealthServer) IsHealthy() (bool, time.Time, time.Time) {
```
We are exporting a method that outputs two `time.Time` var that are never used and are really obscure.
I suggest to have an internal variable with the healthy status and use IsHealthy() to return its value only, keeping this logic where it was. Same as you did with the services_health
However, wait for @danwinship and @thockin opinions on this

alexanderConstantinescu replied:
Yes, that makes more sense. I will change

aojea commented on pkg/proxy/healthcheck/healthcheck_test.go:
```diff
@@ -127,8 +127,10 @@ type healthzPayload struct {
 func TestServer(t *testing.T) {
 	listener := newFakeListener()
 	httpFactory := newFakeHTTPServerFactory()
+	fakeClock := testingclock.NewFakeClock(time.Now())
```
We have to add new testcases with the new combinations that we are adding

alexanderConstantinescu replied:
Will do

alexanderConstantinescu commented:
> We've found recently that the node handlers are feature gated, if we want to use this handlers we should remove the feature gate check on the Node informer [#111344 (comment)](https://github.com/kubernetes/kubernetes/pull/111344#discussion_r936833738)

Hmm, I would have imagined the node handler to be feature gated because `TopologyAwareHints` is, not because the node handler specifically is? If the latter is true, what does that even mean for the node handler to be feature gated - it's not a feature in itself, but more like a means to implementing a feature - like `TopologyAwareHints`.  

In any case: we could also just dial the Kubelet's healthz port on `localhost`, that should be equivalent to listing / watching the node object and inspecting if its `NodeReady` condition is `ConditionTrue`. The problem would be trying to work out the taints the node object has. But I would say the taint we currently look at (`ToBeDeletedByClusterAutoscaler`) is just pre-mature optimization for a node which will be deleted or go `NotReady` in the future. That taint doesn't mean anything _right now_, and if it really is going to be deleted in the near future, then we could just start flagging the node as "not serviceable for LB traffic" when that happens.  PR #105946 didn't really go a very good job at mentioning why we do this to begin with:

> Prior to https://github.com/kubernetes/kubernetes/pull/90823, these nodes would get removed from the LoadBalancer backend once they were marked as Unschedulable. Now that Unschedulable nodes are tolerated by the node predicate, these nodes will only get removed when the nodes have been deleted. If multiple such nodes are being deleted, the list of nodes passed to cloud-provider code via UpdateLoadBalancer() will be stale, causing a few failed updates before the backends stabilize.

thockin commented:
I was away for a bit but I spent all afternoon thinking about this topic and I am less confident than ever :)

An even more revised drawing (now v3):
![Node_LB_states_v3](https://user-images.githubusercontent.com/5595220/191383372-57cdc1f6-4be0-4f43-abcf-7bf60a8d2588.png)

A few things of note:

### 1) How to handle `deletionTimestamp` being set

a) We could exclude being-deleted nodes as a predicate (e.g. `notBeingDeletedPredicate`).  This is simple to do here, and well-behaved clusters already drained the node anyway (with `unschedulable`).  We might miss out on some extra draining time while being deleted, but that seems pretty minor (I don't think cluster-autoscaler uses finalizers).

b) Or we could NOT exclude them (but comment in code why not), and make the on-node HC fail while being deleted.  This gives us extra time to drain LB traffic, but well-behaved clusters already drained the node anyway (with `unschedulable`).

I think (a).

### How to handle the `ToBeDeletedByClusterAutoScaler` taint

See https://github.com/kubernetes/autoscaler/issues/5201 for more context

a) We could exclude tained nodes (as we do today).  The way cluster-autosclaer seems to work means that down-scaled nodes would get `unschedulable` and `ToBeDeletedByClusterAutoScaler` set at the same time, and thus get no LB drain time (as we do today).

b) We could NOT exclude them (but comment in code why not) and maybe make the on-node HC fail.  I think this will break autoscalers who try to delete the underlying VM before the k8s node.  Perhaps we should use finalizers (clear it when all LBs have been updated), which would add weight to (1a) above.

I think (a) for now, but I find it clumsy, and hope I am just mis-reading.

### How to handle `unschedulable` being true

We already do not handle this in the controller.

a) We could ignore it (as we do today).  Given that `unschedulable` often means a node is going away "soon", this would let such nodes handle new connections right up until it is deleted, which cause abrupt errors.

b) We could make all LB->Node HCs fail when `unschedulable` is set.  This would drain LB traffic on some clouds (no new connections, retain existing ones), which makes sense if you consider that `unschedulable` often means a node is going away "soon".   For `eTP: Local` services, this would mean any endpoints on that node are no longer available to serve, even during draining.  If you have a one-replica service and start draining the node that lone pod is on, the service will be unavailable until a new pod is started elsewhere.

c) We could make `eTP: Cluster` LB->Node HCs fail when `unschedulable` is set, while `eTP: Local` would pass.  This would drain "at-large" LB traffic on some clouds (no new connections, retain existing ones), which makes sense if you consider that `unschedulable` often means a node is going away "soon", but would let `eTP: Local` services continue to serve.  Presumable such pods would be gracefully terminated and restarted elsewhere.

I think (c) makes sense, but it a bit more subtle than we discussed before.

What do you think?

alexanderConstantinescu commented:
/remove-lifecycle rotten
/hold cancel
/assign @aojea

alexanderConstantinescu commented:
Update:

On https://github.com/kubernetes/enhancements/pull/3837 we agreed that the fact that Kube-proxy does not consider its healthz state when answering to the HC for eTP:Local services, to be a bug. 

I've updated this PR to only address that bug. 

I will file an issue for this and link it to this bug.

alexanderConstantinescu commented:
/retitle [Proxy]: add `healthz` verification when determining HC response for eTP:Local

danwinship commented on pkg/proxy/healthcheck/service_health.go:
```diff
@@ -226,14 +229,15 @@ func (h hcHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
-	if count == 0 {
-		resp.WriteHeader(http.StatusServiceUnavailable)
-	} else {
+	if count != 0 && healthzHealthy {
```
You should probably update the JSON body as well. eg, add a `"kubeProxyHealthy": %v`, to make it clear why it's failing in this case

alexanderConstantinescu replied:
Updated!

alexanderConstantinescu commented on pkg/proxy/healthcheck/proxier_health.go:
```diff
@@ -39,6 +39,12 @@ type ProxierHealthUpdater interface {
+	// performance w.r.t updating the service rules on the node, specifically:
+	// it's healthy when it's fully synced and not timing out syncing the
+	// service rules.
+	IsHealthy() bool
```
OK, I changed the wording

thockin commented on pkg/proxy/healthcheck/service_health.go:
```diff
@@ -52,7 +52,7 @@ type ServiceHealthServer interface {
 }
 
-func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string) ServiceHealthServer {
+func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
```
I would maybe do:

```
type proxyHealthChecker interface {
  IsHealthy() bool
}

func newServiceHealthServer(..., proxyChecker proxyHealthChecker) ServiceHealthServer {
```

alexanderConstantinescu replied:
We could, [but every proxier implementation is initialized with a `ProxierHealthUpdater`](https://github.com/kubernetes/kubernetes/blob/f286545c48fa04f609cb923ac39b5c7679098054/pkg/proxy/iptables/proxier.go#L237), so we would need to do wire things like:

```go
// ProxierHealthUpdater allows callers to update healthz timestamp only.
type ProxierHealthUpdater interface {
	// QueuedUpdate should be called when the proxier receives a Service or Endpoints
	// event containing information that requires updating service rules.
	QueuedUpdate()

	// Updated should be called when the proxier has successfully updated the service
	// rules to reflect the current state.
	Updated()

	// Run starts the healthz HTTP server and blocks until it exits.
	Run() error

	proxierHealthChecker
}

type proxierHealthChecker interface {
	// IsHealthy returns the proxier's health state, following the same
	// definition the HTTP server defines.
	IsHealthy() bool
}
```

maybe I'm missing something?

thockin replied:
I just meant an interface to decouple and maybe make tesing easier.

I checked out your PR and made this change:

```diff
diff --git a/pkg/proxy/healthcheck/service_health.go b/pkg/proxy/healthcheck/service_health.go
index 277a127d9dc..44f3d255afe 100644
--- a/pkg/proxy/healthcheck/service_health.go
+++ b/pkg/proxy/healthcheck/service_health.go
@@ -52,7 +52,12 @@ type ServiceHealthServer interface {
 	SyncEndpoints(newEndpoints map[types.NamespacedName]int) error
 }
 
-func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
+// healthChecker is something that can report whether it is healthy or not.
+type healthChecker interface {
+	IsHealthy() bool
+}
+
+func newServiceHealthServer(hostname string, recorder events.EventRecorder, listener listener, factory httpServerFactory, nodePortAddresses []string, proxyChecker healthChecker) ServiceHealthServer {
 
 	nodeAddresses, err := utilproxy.GetNodeAddresses(nodePortAddresses, utilproxy.RealNetwork{})
 	if err != nil || nodeAddresses.Len() == 0 {
@@ -75,15 +80,15 @@ func newServiceHealthServer(hostname string, recorder events.EventRecorder, list
 		recorder:      recorder,
 		listener:      listener,
 		httpFactory:   factory,
-		healthzServer: healthzServer,
+		proxyChecker:  proxyChecker,
 		services:      map[types.NamespacedName]*hcInstance{},
 		nodeAddresses: nodeAddresses,
 	}
 }
 
 // NewServiceHealthServer allocates a new service healthcheck server manager
-func NewServiceHealthServer(hostname string, recorder events.EventRecorder, nodePortAddresses []string, healthzServer ProxierHealthUpdater) ServiceHealthServer {
-	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses, healthzServer)
+func NewServiceHealthServer(hostname string, recorder events.EventRecorder, nodePortAddresses []string, proxyChecker healthChecker) ServiceHealthServer {
+	return newServiceHealthServer(hostname, recorder, stdNetListener{}, stdHTTPServerFactory{}, nodePortAddresses, proxyChecker)
 }
 
 type server struct {
@@ -94,7 +99,7 @@ type server struct {
 	listener      listener
 	httpFactory   httpServerFactory
 
-	healthzServer ProxierHealthUpdater
+	proxyChecker healthChecker
 
 	lock     sync.RWMutex
 	services map[types.NamespacedName]*hcInstance
@@ -229,7 +234,7 @@ func (h hcHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
 		return
 	}
 	count := svc.endpoints
-	kubeProxyHealthy := h.hcs.healthzServer.IsHealthy()
+	kubeProxyHealthy := h.hcs.proxyChecker.IsHealthy()
 	h.hcs.lock.RUnlock()
 
 	resp.Header().Set("Content-Type", "application/json")
```

That compiles.

alexanderConstantinescu replied:
OK, updated the PR with this

thockin reviewed (commented):
looks good, just needs rebase. minor comment

aojea commented on pkg/proxy/healthcheck/proxier_health.go:
```diff
@@ -94,6 +98,37 @@ func (hs *proxierHealthServer) QueuedUpdate() {
+	return isHealthy
+}
+
+func (hs *proxierHealthServer) isHealthy() (bool, time.Time, time.Time) {
```
just a thought, we are returning the two `time.Time` just for the output message, why if we just output the message directly?

```
// isHealthy returns the state of the proxy and a message with the reason in case is not healthy
func (hs *proxierHealthServer) isHealthy() (bool, string)
...
return false, fmt.Printf(`{"lastUpdated": %q,"currentTime": %q}`, lastUpdated, currentTime)
```

or just an error with the message

```
// isHealthy returns and error if not healthy
func (hs *proxierHealthServer) isHealthy() error
...
return fmt.Errorf(`{"lastUpdated": %q,"currentTime": %q}`, lastUpdated, currentTime)
```

alexanderConstantinescu replied:
I will let owners / people more experienced with the proxy code comment on this.

thockin replied:
I would not rely on an `error.Error()` to return valid JSON - that's a little weird.

We could return the JSON as `[]byte` or even return a struct but I don't think it's worth the extra plumbing.

aojea commented on pkg/proxy/healthcheck/proxier_health.go:
```diff
@@ -94,6 +98,37 @@ func (hs *proxierHealthServer) QueuedUpdate() {
+		oldestPendingQueued = val.(time.Time)
+	}
+	if val := hs.lastUpdated.Load(); val != nil {
+		lastUpdated = val.(time.Time)
```
@danwinship why do we need lastUpdated, it doesn't seem to be used anywhere?

alexanderConstantinescu replied:
Maybe: https://github.com/kubernetes/kubernetes/blob/f286545c48fa04f609cb923ac39b5c7679098054/pkg/proxy/healthcheck/proxier_health.go#L168-L173

?

aojea replied:
yeah, it is always set to currentTime , no? why do we need to get this value, is not used anywhere AFAIK

aojea replied:
I mean, we can return `oldestPendingQueued` instead of `lastUpdated`

aojea replied:
nevermind, it doesn't solve anything, I just wanted to see if we can save a variable and I said nonsense, please resolve this comment

alexanderConstantinescu committed:
Add proxy healthz result to ETP=local health check

Today, the health check response to the load balancers asking Kube-proxy for
the status of ETP:Local services does not include the healthz state of Kube-
proxy. This means that Kube-proxy might indicate to load balancers that they
should forward traffic to the node in question, simply because the endpoint
is running on the node - this overlooks the fact that Kube-proxy might be
not-healthy and hasn't successfully written the rules enabling traffic to
reach the endpoint.

thockin commented on pkg/proxy/healthcheck/healthcheck_test.go:
```diff
@@ -349,9 +352,29 @@ func TestServer(t *testing.T) {
+	tHanlder(hcs, nsn, status, endpoints, kubeProxyHealthy, t)
+}
+
+func tHanlder(hcs *server, nsn types.NamespacedName, status int, endpoints int, kubeProxyHealthy bool, t *testing.T) {
```
tHandler - spelling :)

thockin commented on pkg/proxy/healthcheck/healthcheck_test.go:
```diff
@@ -131,8 +132,10 @@ type healthzPayload struct {
 	listener := newFakeListener()
 	httpFactory := newFakeHTTPServerFactory()
+	fakeClock := testingclock.NewFakeClock(time.Now())
+	hs := newProxierHealthServer(listener, httpFactory, fakeClock, "127.0.0.1:10256", 10*time.Second, nil, nil)
```
The `interface` in my other comment means you don't have to do this fake-clock stuff (which I find hard to reason about and is not germane to this test)

```diff
diff --git a/pkg/proxy/healthcheck/healthcheck_test.go b/pkg/proxy/healthcheck/healthcheck_test.go
index dd3e9b76976..8db3feaecfb 100644
--- a/pkg/proxy/healthcheck/healthcheck_test.go
+++ b/pkg/proxy/healthcheck/healthcheck_test.go
@@ -129,13 +129,20 @@ type healthzPayload struct {
 	CurrentTime string
 }
 
+type fakeHealthChecker struct {
+	value bool
+}
+
+func (fake fakeHealthChecker) IsHealthy() bool {
+	return fake.value
+}
+
 func TestServer(t *testing.T) {
 	listener := newFakeListener()
 	httpFactory := newFakeHTTPServerFactory()
-	fakeClock := testingclock.NewFakeClock(time.Now())
-	hs := newProxierHealthServer(listener, httpFactory, fakeClock, "127.0.0.1:10256", 10*time.Second, nil, nil)
+	proxyChecker := &fakeHealthChecker{true}
 
-	hcsi := newServiceHealthServer("hostname", nil, listener, httpFactory, []string{}, hs)
+	hcsi := newServiceHealthServer("hostname", nil, listener, httpFactory, []string{}, proxyChecker)
 	hcs := hcsi.(*server)
 	if len(hcs.services) != 0 {
 		t.Errorf("expected 0 services, got %d", len(hcs.services))
@@ -354,13 +361,13 @@ func TestServer(t *testing.T) {
 	testHandler(hcs, nsn4, http.StatusOK, 6, t)
 
 	// fake a temporary unhealthy proxy
-	hs.oldestPendingQueued.Store(hs.clock.Now().Add(-hs.healthTimeout - time.Second))
+	proxyChecker.value = false
 	testHandlerWithHealth(hcs, nsn2, http.StatusServiceUnavailable, 0, false, t)
 	testHandlerWithHealth(hcs, nsn3, http.StatusServiceUnavailable, 7, false, t)
 	testHandlerWithHealth(hcs, nsn4, http.StatusServiceUnavailable, 6, false, t)
 
 	// fake a healthy proxy
-	hs.oldestPendingQueued.Store(hs.clock.Now())
+	proxyChecker.value = true
 	testHandlerWithHealth(hcs, nsn2, http.StatusServiceUnavailable, 0, true, t)
 	testHandlerWithHealth(hcs, nsn3, http.StatusOK, 7, true, t)
 	testHandlerWithHealth(hcs, nsn4, http.StatusOK, 6, true, t)
```

thockin commented on pkg/proxy/healthcheck/healthcheck_test.go:
```diff
@@ -434,12 +467,14 @@ func testHealthzHandler(server httpServer, status int, t *testing.T) {
 func TestServerWithSelectiveListeningAddress(t *testing.T) {
 	listener := newFakeListener()
 	httpFactory := newFakeHTTPServerFactory()
+	fakeClock := testingclock.NewFakeClock(time.Now())
```
Any reason not to use the simpler fake here?

alexanderConstantinescu replied:
No other reason other than: I overlooked it

thockin commented on pkg/proxy/healthcheck/proxier_health.go:
```diff
@@ -41,6 +41,14 @@ type ProxierHealthUpdater interface {
+	proxierHealthChecker
+}
+
+type proxierHealthChecker interface {
```
This is not something you need to act on, but I wanted to call out a small distinction.

You can define this interface here, next to the "producer", and it's fine (it's all the same pkg, after all).  Or you can define it next to the consumer, embracing the Go idea of "produce types, accept interfaces".  This file has no use of this interface, but the service_health (and test) file gets simpler because of it.

The receiver is saying "I need something I can ask IsHealthy()", the producer is passing something that satisfies it.

Again, it's fine this way, too.

alexanderConstantinescu replied:
I wasn't aware of these notions. You got me reading up on something :) I updated the PR with this

thockin reviewed (commented):
/lgtm
/approve
/hold


minor comment in test - remove hold if you just want to go

thockin commented on pkg/proxy/healthcheck/service_health.go:
```diff
@@ -226,24 +235,26 @@ func (h hcHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
 			},
-			"localEndpoints": %d
+			"localEndpoints": %d,
+			"kubeProxyHealthy": %v
```
If we expect this to be an API that different proxy implementations could implement, maybe we should call this `dataplaneHealthy` or something?  Specifically not "kubeProxy" ?

alexanderConstantinescu replied:
At this point this isn't library code so I wouldn't consider it an API. That being said: I do know some proxy implementations copy this code, so...sure, I'll update

thockin reviewed (commented):
sorry, 1 more thing

thockin commented:
Thanks!

/lgtm
/approve

thockin commented:
```
=== Failed
2595
=== FAIL: pkg/proxy/healthcheck TestServer (0.11s)
2596
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2597
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2598
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2599
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2600
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2601
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2602
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2603
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2604
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2605
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2606
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2607
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2608
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2609
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2610
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2611
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2612
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2613
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2614
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2615
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2616
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2617
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
2618
2619
=== FAIL: pkg/proxy/healthcheck TestServerWithSelectiveListeningAddress (0.00s)
2620
    healthcheck_test.go:413: expected true kubeProxyHealthy, got false
```

alexanderConstantinescu commented:
Fixed the failing tests and re-pushed

thockin commented:
/lgtm
/retest

alexanderConstantinescu commented:
/hold cancel
